### PR TITLE
feat: feed source

### PR DESCRIPTION
**Pull request by ice-orion**
### What does this PR do?
This PR adds datasource for the feed, implements follow feature

### Changes brought by this PR
* Add `findBestOptions` + `feedDataSourceProvider` to find best relays to fetch on feed
* Add custom exceptions with code and message + some exceptions for specific cases
* Add `followListProvider` + method to toggle "follow"

### CODE REVIEW

File: lib/app/features/user/pages/profile_page/components/profile_details/profile_followers_cell.dart [hunk 1]
```diff
@@ -23,7 +23,7 @@ class ProfileFollowersCell extends ConsumerWidget {
   @override
   Widget build(BuildContext context, WidgetRef ref) {
     final usersNumber = followType == FollowType.following
-        ? ref.watch(userFollowingProvider(pubkey)).valueOrNull?.length ?? 0
+        ? ref.watch(FollowListProvider(pubkey)).valueOrNull?.data.list.length ?? 0
```
Comment (ice-alcides) [inline lib/app/features/user/pages/profile_page/components/profile_details/profile_followers_cell.dart:26]: Need to change with a lower case

**Review by ice-alcides — APPROVED** [review]

File: lib/app/utils/algorithm.dart [hunk 2]
```diff
@@ -0,0 +1,53 @@
...
+///   "Key1": ["Option1", "Option2", "Option3", "Option4"],
+///   "Key2": ["Option4", "Option5", "Option6", "Option7"],
+///   "Key3": ["Option8", "Option9", "Option10", "Option11"]
+/// }
+/// The output will be:
+/// {
+///   "Option4": ["Key1", "Key2"],
+///   "Option8": ["Key3"]
+/// }
+Map<String, List<String>> findBestOptions(Map<String, List<String>> keysToOptions) {
```
Comment (ice-tychon) [inline lib/app/utils/algorithm.dart:17]: nit: Should we add tests for this?

**Review by ice-kreios — DISMISSED** [review]

File: lib/app/utils/algorithm.dart [hunk 3]
```diff
@@ -0,0 +1,53 @@
...
+  /// Transforms
+  /// {
+  ///   "Key1": ["Option1", "Option2"],
+  ///   "Key1": ["Option2", "Option3"]
```
Comment (ice-tychon) [inline lib/app/utils/algorithm.dart:21]: Should be "Key2"

File: lib/app/utils/algorithm.dart [hunk 4]
```diff
@@ -0,0 +1,53 @@
...
+  /// }
+  /// to
+  /// {
+  ///   "Option1": ["Key1"],
+  ///   "Option2": ["Key1", "Key2"],
+  ///   "Option3": ["Key1"]
```
Comment (ice-tychon) [inline lib/app/utils/algorithm.dart:27]: Also here, 
```suggestion
  ///   "Option3": ["Key2"]
```

**Review by ice-alcides — DISMISSED** [review]

File: lib/app/utils/algorithm.dart [hunk 5]
```diff
@@ -0,0 +1,53 @@
...
+///   "Key2": ["Option4", "Option5", "Option6", "Option7"],
+///   "Key3": ["Option8", "Option9", "Option10", "Option11"]
+/// }
+/// The output will be:
+/// {
+///   "Option4": ["Key1", "Key2"],
+///   "Option8": ["Key3"]
+/// }
+Map<String, List<String>> findBestOptions(Map<String, List<String>> keysToOptions) {
+  /// Transforms
```
Comment (ice-tychon) [inline lib/app/utils/algorithm.dart:18]: Add this check, 'cause `reduce` may fail (throw an exception) if the input map is empty.
```suggestion
  if (keysToOptions.isEmpty) {
    return {};
  }
  /// Transforms
```

Or even change `reduce` to `fold` on the 37th line.

File: lib/app/utils/algorithm.dart [hunk 6]
```diff
@@ -0,0 +1,53 @@
...
+    for (final option in entry.value) {
+      optionsToKeys.putIfAbsent(option, () => <String>[]).add(entry.key);
+    }
+  }
+
+  /// Looking for the "best" entry - entry with most keys
+  final maxEntry = optionsToKeys.entries.reduce(
+    (maxEntry, currentEntry) =>
+        currentEntry.value.length > maxEntry.value.length ? currentEntry : maxEntry,
+  );
```
Comment (ice-tychon) [inline lib/app/utils/algorithm.dart:40]:
```suggestion
  final maxEntry = maxBy(optionsToKeys.entries, (entry) => entry.value.length)
```

File: lib/app/exceptions/exceptions.dart [hunk 7]
```diff
@@ -0,0 +1,49 @@
+// SPDX-License-Identifier: ice License 1.0
+
+sealed class IonException implements Exception {
```
Comment (ice-tychon) [inline lib/app/exceptions/exceptions.dart:3]: Rename to `IONException`

**Review by ice-alcides — DISMISSED** [review]

**Review by ice-alcides — APPROVED** [review]

**Review by ice-tychon — APPROVED** [review]

**Pull request merged**